cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,192 @@
+/// ```text
+/// jj log
+/// ◉   3
+/// ◉   2
+/// ◉   1
+/// ◉
+/// jj parallelize 1::2
+/// jj log
+/// ◉     3
+/// ├─╮
+/// │ ◉   2
+/// ◉ │   1
+/// ├─╯
+/// ◉
+/// ```
+#[derive(clap::Args, Clone, Debug)]
+#[command(verbatim_doc_comment, visible_aliases=&["para"])]
+pub(crate) struct ParallelizeArgs {
+    revisions: Vec<RevisionArg>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_parallelize(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &ParallelizeArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let mut tx = workspace_command.start_transaction();
+    // The target commits are the commits being parallelized.
+    let target_commits: Vec<CommitId> = tx
+        .base_workspace_helper()
+        .parse_union_revsets(&args.revisions)?
+        .evaluate()?
+        .iter()
+        // We want parents before children, so we need to reverse the order.
+        .reversed()
+        .collect();
+    if target_commits.len() < 2 {
+        return Ok(writeln!(ui.status(), "Nothing changed.")?);
+    }
+    let target_revset = RevsetExpression::commits(target_commits.clone());
+    let target_head = get_head(&target_revset, tx.repo())?;
+    let target_root = get_root(&target_revset, tx.repo())?;
+    let connected_length = RevsetExpression::commits(vec![target_head, target_root.clone()])
+        .connected()
+        .evaluate_programmatic(tx.repo())?
+        .iter()
+        .count();
+    if connected_length != target_commits.len() {
+        return Err(user_error(
+            "Cannot parallelize since the target revisions are not connected.",
+        ));
+    }
+
+    // Rebase the children of each target commit onto its new parents. A child
+    // which had a target commit as an ancestor before parallelize ran will have
+    // the target commit as a parent afterward.
+    for target_commit in target_commits.iter() {
+        // These parents are shared by all children of the target commit and
+        // include the target commit itself plus any of its ancestors which are
+        // being parallelized.
+        let common_parents: IndexSet<CommitId> = target_revset
+            .intersection(&RevsetExpression::commit(target_commit.clone()).ancestors())
+            .evaluate_programmatic(tx.repo())?
+            .iter()
+            // We want parents before children, so we need to reverse the order.
+            .reversed()
+            .collect();
+        // Children of the target commit, excluding other target commits.
+        let children: Vec<Commit> = RevsetExpression::commit(target_commit.clone())
+            .children()
+            .minus(&target_revset)
+            .evaluate_programmatic(tx.repo())?
+            .iter()
+            .commits(tx.repo().store())
+            .try_collect()?;
+        for child in children {
+            let mut new_parents = common_parents.clone();
+            new_parents.extend(child.parent_ids().iter().cloned());
```

Comment:
nit: can you add a test that would fail if we didn't do `.extend()`?

fwiw, it might be actually okay to error out if inner revisions had children not included in the target revisions. I don't think the user would want to create a merge commit except for the one from the target head's children.

Reply:
I think it makes logical sense to preserve the dependencies by making any descendants remain descendants of the same commits as before. The user can always rebase the commit somewhere else afterwards, or they can undo if they don't like it. OTOH, I think it's going to be a very rare thing to have this shape before in the first place, so you may be right that it's usually a mistake. I'm fine either way.

Reply:
I added a new test called `test_parallelize_with_merge_commit_child`. I'm not sure if we should error out here or not. It certainly seems odd for someone to call parallelize in this situation, but I'm not sure if we should stop them if they want to.

This case isn't that different than the one in `test_parallelize_with_descendants_complex` (they both create merges). So if we block this one, maybe we should block that one as well? I think that would mean that parallelize only works if the only the head commit has children that aren't being parallelized (i.e. the commits are in a straight line). @martinvonz do you have any opinions about this?

Reply:
Isn't `test_parallelize_with_descendants_complex` exactly the case Yuya was asking about? Which other case is there? Do you mean `test_parallelize_with_descendants_simple`? I think that case is going to be pretty common.

Anyway, I'm leaning towards your "I'm not sure if we should stop them if they want to"

Reply:
I just thought it would be rare enough to optimize the children lookup out of the loop. Maybe there will be a better way to address that, though.

Reply:
I thought Yuya was asking about a case where the child of a target commit is a merge commit. In `test_parallelize_with_descendants_complex` the child is not a merge commit.